### DISCUSSION TITLE
Weird behaviour using BVH with small amount of objects

### DISCUSSION THREAD

**Discussion by niccolot** [RayTracing/raytracing.github.io#1646] [Debugging]
Hello, i' ve so far implemented the raytracer  as explained in the books. When testing the BVH i' ve seen the performance boost with a large image as expected. When trying it with a small test image not only using BVH takes significantly longer (which i think is expected since i' ve put the 3 sphere in line so no well separable) but the 2 images are different in both appearance of the materials and position of the blue sphere:

BVH off:
![bvh_off](https://github.com/user-attachments/assets/3241571b-7712-4c4c-b469-c1e9356d4f66)

BVH on:
![bvh_on](https://github.com/user-attachments/assets/fa52a083-ae22-45c2-9cdc-a7aea66f7c6b)

The inciminated main is

```c++
#include <iostream>
#include <memory>

#include "camera.h"
#include "sphere.h"
#include "material.h"
#include "bvh.h"

int main() {
    
    int img_width = 512;
    double aspect_ratio = 16.0 / 9.0;

    HittableList world;

    auto Material_ground = std::make_shared<Lambertian>(Color(0.8, 0.8, 0.0));
    auto Material_center = std::make_shared<Lambertian>(Color(0.1, 0.2, 0.5));
    auto Material_left   = std::make_shared<Dielectric>(1.5);
    auto Material_right  = std::make_shared<Metal>(Color(0.8, 0.6, 0.2), 0.3);

    auto center1 = Vec3(0.0, 0.0, -1.2);
    auto center2 = center1 + Vec3(0.0, 0.25, -1.2);

    world.add(std::make_shared<Sphere>(Vec3( 0.0, -100.5, -1.0), 100.0, Material_ground));
    world.add(std::make_shared<Sphere>(center1, center2, 0.5, Material_center));
    world.add(std::make_shared<Sphere>(Vec3(-1.0,    0.0, -1.0),   0.5, Material_left));
    world.add(std::make_shared<Sphere>(Vec3( 1.0,    0.0, -1.0),   0.5, Material_right));

    Camera cam(img_width, aspect_ratio, Vec3(0,0,1), Vec3(0,0,.99), 45);

    world = HittableList(std::make_shared<BVHNode>(world));

    cam.render(world);
}
```

Everything is reproducible by just running the build.sh script in my  [repo](https://github.com/niccolot/path_tracer) and commenting/uncommenting the BVH line

**Comment by hollasch** [comment]
It may help to check out branch `progression/2/2024-0405`. This is the latest full progression branch for book two. From this branch, see commit 21990cb2, which is the state of the source after listing 2-21. You can use this code as a reference implementation, and sub in your `main()` procedure above. Does this demonstrate the same difference in behavior?

**Reply by niccolot** [reply]
Using your branch i don't see any difference (and also not a dramatic decrease in speed with bvh on and small number of lined objects as with my code):

bvh off:
![no_bvh](https://github.com/user-attachments/assets/af5c3f43-85bf-4fa1-a2ba-3c2bc89216bd)

bvh on:
![bvh](https://github.com/user-attachments/assets/e4510fe3-bda4-4c0f-9e76-42f95ca791ec)

the main i used is 

```c++
int main() {

    hittable_list world;

    auto Material_ground = std::make_shared<lambertian>(color(0.8, 0.8, 0.0));
    auto Material_center = std::make_shared<lambertian>(color(0.1, 0.2, 0.5));
    auto Material_left   = std::make_shared<dielectric>(1.5);
    auto Material_right  = std::make_shared<metal>(color(0.8, 0.6, 0.2), 0.3);

    auto center1 = vec3(0.0, 0.0, -1.);
    auto center2 = center1 + vec3(0.0, 0.25, 0);

    world.add(std::make_shared<sphere>(vec3( 0.0, -100.5, -1.0), 100.0, Material_ground));
    world.add(std::make_shared<sphere>(center1, center2, 0.5, Material_center));
    world.add(std::make_shared<sphere>(vec3(-1.0,    0.0, -1.0),   0.5, Material_left));
    world.add(std::make_shared<sphere>(vec3( 1.0,    0.0, -1.0),   0.5, Material_right));

    camera cam;
    cam.aspect_ratio      = 16.0 / 9.0;
    cam.image_width       = 400;
    cam.samples_per_pixel = 100;
    cam.max_depth         = 50;
    cam.background        = color(0.70, 0.80, 1.00);

    cam.vfov     = 90;
    cam.lookfrom = point3(0,0,0.25);
    cam.lookat   = point3(0,0,-1);
    cam.vup      = vec3(0,1,0);

    cam.defocus_angle = 0.;
    cam.focus_dist    = 1.0;
    

    world = hittable_list(std::make_shared<bvh_node>(world));

    cam.render(world);
}
```

i think its worth for me going through each part of the bvh implementation in case i missed something, thank you for your help so far

**Reply by hollasch** [reply]
Well, can't say I'm not relieved. Hope this helps you quickly spot the issue in your code.

**Reply by niccolot** [reply]
The magic of programming struck again. While i was tidying up some parts it came to my mind to check again and now it's all good so there is a 50/50 chance that i corrected the error by chance or was straight up misusing the camera in the previous example (looking through the whole code it seems to be ok). Anyway many thanks for your attention and for the project in general!

**Comment by niccolot** [comment]
All good now

**Reply by hollasch** [reply]
Excellent!
